sdk/program/src/vote/state/vote_state_deserialize.rs

```diff
@@ -0,0 +1,150 @@
+use {
+    crate::{
+        instruction::InstructionError,
+        pubkey::Pubkey,
+        serialize_utils::cursor::*,
+        vote::state::{BlockTimestamp, LandedVote, Lockout, VoteState, MAX_ITEMS},
+    },
+    bincode::serialized_size,
+    std::io::Cursor,
+};
+
+pub(super) fn deserialize_vote_state_into(
+    input: &[u8],
+    cursor: &mut Cursor<&[u8]>,
+    vote_state: &mut VoteState,
+    has_latency: bool,
+) -> Result<(), InstructionError> {
+    vote_state.node_pubkey = read_pubkey(cursor)?;
+    vote_state.authorized_withdrawer = read_pubkey(cursor)?;
+    vote_state.commission = read_u8(cursor)?;
+    read_votes_into(cursor, vote_state, has_latency)?;
+    vote_state.root_slot = read_option_u64(cursor)?;
+    read_authorized_voters_into(cursor, vote_state)?;
+
+    // `serialized_size()` must be used over `mem::size_of()` because of alignment
+    let position_after_prior_voters = cursor
+        .position()
+        .checked_add(
+            serialized_size(&vote_state.prior_voters)
+                .map_err(|_| InstructionError::InvalidAccountData)?,
+        )
+        .ok_or(InstructionError::InvalidAccountData)?;
+
+    let is_empty_idx = position_after_prior_voters
+        .checked_sub(1)
+        .ok_or(InstructionError::InvalidAccountData)?;
+
+    match input[is_empty_idx as usize] {
+        0 => read_prior_voters_into(cursor, vote_state)?,
+        1 => cursor.set_position(position_after_prior_voters),
+        _ => return Err(InstructionError::InvalidAccountData),
+    }
```

Comment:
nit: It's a bit strange and potentially error-prone to take in the input bytes and also a cursor over the same input bytes.

Although it might make the code slightly uglier, I'd prefer to see it all operate on the cursor, so this would need to change to move the cursor to `is_empty_idx`, read the u8 there, and optionally move it back to the previous position if it's not empty. Does that sound reasonable?

Reply:
we could do that. honestly we could also just remove this logic and have it step over the bytes normally, it turns out to not be that much of an optimization since the final version of `read_prior_voters_into` doesnt memcpy the null pubkeys

Reply:
went ahead and did that